cli/explain/compactgraph/compactgraph.go

```diff
@@ -0,0 +1,478 @@
+					otherDiff.TransitivelyInvalidated = make(map[string]uint32)
+				}
+				for k, v := range result.spawnDiff.TransitivelyInvalidated {
+					otherDiff.TransitivelyInvalidated[k] += v
+				}
+				otherDiff.TransitivelyInvalidated[spawn.Mnemonic]++
+			}
+		}
+		if len(result.spawnDiff.Diffs) > 0 && (result.localChange || !foundTransitiveCause) {
+			spawnDiffs = append(spawnDiffs, result.spawnDiff)
+		}
+	}
+
+	return
+}
+
+// ReduceToRoots returns the sorted subset of outputs that are not inputs to any other spawn in the graph.
+func (cg *CompactGraph) ReduceToRoots(outputs []string) []string {
+	rootsSet := make(map[string]struct{})
+	for _, output := range outputs {
+		rootsSet[output] = struct{}{}
+	}
+
+	// Visit all nodes in the graph and remove those with incoming edges from the set of roots.
+	var toVisit []interface{}
+	visited := make(map[interface{}]struct{})
+	markForVisit := func(n interface{}) {
+		if _, seen := visited[n]; !seen {
+			toVisit = append(toVisit, n)
+		}
+	}
+	for _, root := range outputs {
+		markForVisit((*cg)[root])
+	}
+	for len(toVisit) > 0 {
+		var node interface{}
+		node, toVisit = toVisit[0], toVisit[1:]
+		visited[node] = struct{}{}
+		switch n := node.(type) {
+		case *File:
+		case *Directory:
+			delete(rootsSet, n.Path())
+		}
+		cg.visitSuccessors(node, func(input interface{}) {
+			markForVisit(input)
+		})
+	}
+
+	roots := maps.Keys(rootsSet)
+	slices.Sort(roots)
+	return roots
+}
+
+// SortTopologically returns the primary outputs of the graph in topological order.
+func (cg *CompactGraph) SortTopologically() []string {
+	var toVisit []interface{}
+	for _, spawn := range *cg {
+		toVisit = append(toVisit, spawn)
+	}
+	sort.Slice(toVisit, func(i, j int) bool {
+		return toVisit[i].(*Spawn).PrimaryOutputPath() < toVisit[j].(*Spawn).PrimaryOutputPath()
+	})
+
+	ordered := make([]string, 0, len(*cg))
+	state := make(map[interface{}]bool)
+	for len(toVisit) > 0 {
+		n := toVisit[len(toVisit)-1]
+		toVisit = toVisit[:len(toVisit)-1]
+		if done, seen := state[n]; seen {
+			if !done {
+				state[n] = true
+				if spawn, ok := n.(*Spawn); ok {
+					ordered = append(ordered, spawn.PrimaryOutputPath())
+				}
+			}
+			continue
+		}
+		state[n] = false
+		toVisit = append(toVisit, n)
+		cg.visitSuccessors(n, func(input interface{}) {
+			toVisit = append(toVisit, input)
+		})
+	}
+	slices.Reverse(ordered)
+	return ordered
+}
+
+func (cg *CompactGraph) visitSuccessors(node interface{}, visitor func(input interface{})) {
+	switch n := node.(type) {
+	case *File:
+	case *Directory:
+		if !isSourcePath(n.Path()) {
+			spawn, ok := (*cg)[n.Path()]
+			if ok {
+				visitor(spawn)
+			}
+		}
+	case *InputSet:
+		for _, file := range n.Files {
+			visitor(file)
+		}
+		for _, dir := range n.Directories {
+			visitor(dir)
+		}
+		for _, inputSet := range n.TransitiveSets {
+			visitor(inputSet)
+		}
+	case *Spawn:
+		visitor(n.Tools)
+		visitor(n.Inputs)
+	}
+}
+
+func (cg *CompactGraph) primaryOutputs() []string {
```

Comment:
We could, but as we only access this data once, storing it would also be a waste (and complicate `CompactGraph` since we would now be storing the result of a pure function of its other contents). Since we are iterating over the graph many times anyway, I wouldn't expect this to have meaningful CPU time impact.